(#792, #802) Adding ApplicationLauncher to manage app services, including GobblinMetrics lifecyle

sahilTakiar opened the pull request:
#### Issues

* Partial fix for #792
* Full fix for #802

#### Description

* Adding an `ApplicationLauncher` interface that is responsible for launching Gobblin applications
    * Gobblin Applications are launched in a variety of ways - `AzkabanJobLauncher`, `GobblinApplicationMaster`, `CliLocalJobLauncher`, `CliMRJobLauncher`, `SchedulerDaemon`
* A base implementation of `ApplicationLauncher` called `ServiceBasedApplicationLauncher` defines a `ApplicationLauncher` is a collection of Guava `Service`s that need to be run
    * A core set of `Service`s are always run, and sub-classes can add more `Service`s
* All Gobblin Application Launchers mentioned above now extend `ServiceBasedApplicationLauncher`
    * A lot of common functionality between all these classes has been moved into `ServiceBasedApplicationLauncher`
* A `MetricsReportingService` has been added that starts and stops metrics reporting - this fixes the two issues mentioned above since metrics will only be started and stopped once per application
* The `gobblin-scheduler` submodule has been deleted and its classed moved into `gobblin-runtime` - this way `gobblin-runtime` can use the scheduler classes in a more generic manner
* Added a `JMXReportingService` for collecting JVM stats and reporting them via JMX (this code was duplicated in a few places so I moved it into a separate class)

#### Reviewing this PR

* `ApplicationLauncher` and `ServiceBasedApplicationLauncher` are the best places to start

kadaan commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java:
```diff
@@ -28,10 +27,20 @@
 import com.google.common.collect.Lists;
 import com.google.common.io.Closer;
 
+import org.apache.hadoop.conf.Configuration;
```
import ordering: org comes before com

kadaan commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java:
```diff
@@ -125,12 +144,41 @@ public void run()
 
+  @Override
+  public void start() throws ApplicationException {
+    this.applicationLauncher.start();
```
Should we add listener events here?  Maybe as an enhancement.

sahilTakiar replied:
You mean add `ApplicationListener`s to the `start()` and `stop()` method of `ApplicationLauncher`?

kadaan replied:
That was what I was wondering.  It could be useful for users to be able to setup services that need to  start/stop when the application start/stops.

sahilTakiar replied:
Yeah, but the `ServiceBasedAppLauncher.addService` method should be able to handle that. Although it may still be useful for other things so I will still add it.

kadaan replied:
The benefit is that JobListeners can be added just by specifying them in the config files.

kadaan commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java:
```diff
@@ -125,12 +144,41 @@ public void run()
+
+  @Override
+  public void stop() throws ApplicationException {
+    this.applicationLauncher.stop();
```
The same as above.  If this was done, job developers could also have resources that are initialized once at job start and cleaned up at job end.

kadaan commented on gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java:
```diff
@@ -204,10 +202,6 @@ public void apply(JobListener jobListener, JobContext jobContext) throws Excepti
 
   @Override
   public void launchJob(JobListener jobListener) throws JobException {
-    if (this.jobContext.getJobMetricsOptional().isPresent()) {
```
Annoyingly, these changes are going to cause #764 to not merge correctly.  Would be really nice to get #764 in as it was fixing a couple of other bugs.

kadaan commented on gobblin-runtime/src/main/java/gobblin/runtime/local/CliLocalJobLauncher.java:
```diff
@@ -12,27 +12,86 @@
+
+import javax.annotation.Nullable;
+
+import com.google.common.io.Closer;
```
Import ordering

kadaan commented on gobblin-runtime/src/main/java/gobblin/runtime/local/CliLocalJobLauncher.java:
```diff
@@ -12,27 +12,86 @@
+
+  @Override
+  public void close() throws IOException {
+    this.close();
```
This seems to be calling the wrong thing.  Shouldn't this be `this.closer.close()`?

kadaan commented on gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/CliMRJobLauncher.java:
```diff
@@ -12,20 +12,28 @@
+
+import javax.annotation.Nullable;
+
+import com.google.common.io.Closer;
```
Import ordering

kadaan commented on gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java:
```diff
@@ -0,0 +1,203 @@
+    LOG.info("Shutting down the application");
+    try {
+      // Give the services 5 seconds to stop to ensure that we are responsive to shutdown requests
+      this.serviceManager.stopAsync().awaitStopped(5, TimeUnit.SECONDS);
```
5 seconds seems really short.  The `GobblinApplicationMaster` used to wait for 5 minutes.

sahilTakiar replied:
5 minutes seems excessive, I'm going to change it to 1 minute

kadaan replied:
Can this be configurable?  Either in the config file or by the job launcher?

sahilTakiar replied:
Yeah I can make this configurable, with a default of 1 minute

kadaan commented on gobblin-yarn/src/main/java/gobblin/yarn/GobblinApplicationMaster.java:
```diff
@@ -202,17 +166,12 @@ public synchronized void stop() {
-      this.serviceManager.stopAsync().awaitStopped(5, TimeUnit.MINUTES);
-    } catch (TimeoutException te) {
-      LOGGER.error("Timeout in stopping the service manager", te);
+      super.stop();
```
Doesn't this mean that `GobblinApplicationMaster` will only wait 5 seconds for shutdown now, rather than 5 minutes?

kadaan commented on gobblin-yarn/src/main/java/gobblin/yarn/GobblinWorkUnitRunner.java:
```diff
@@ -174,6 +162,7 @@ public GobblinWorkUnitRunner(String applicationName, String helixInstanceName, C
       LOGGER.info("Adding YarnContainerSecurityManager since login is keytab based");
       services.add(new YarnContainerSecurityManager(config, fs, this.eventBus));
     }
+    services.add(new JMXReportingService());
```
Why don't we use `ServiceBasedAppLauncher` for the services in the container as well?

sahilTakiar replied:
We could be there are services like the Job History Store and Admin UI that aren't really needed. Also the Metrics Service would be different as `GobblinWorkUnitRunner` uses `ContainerMetrics`

kadaan commented:
Am I correct that this fixes 1 and 2 of #792 as well as fixing #785?

kadaan commented on gobblin-runtime/src/main/java/gobblin/runtime/services/MetricsReportingService.java:
```diff
@@ -0,0 +1,44 @@
+
+  @Override
+  protected void startUp() throws Exception {
+    GobblinMetrics.get(this.appId).startMetricReporting(this.properties);
```
How is this going to work with the `buildFileMetricReporter()` call made from with `startMetricReporting`?  That method constructs an outputstream which is specific for the job_id that is running.

sahilTakiar replied:
It should just be specific to `this.appId`. The `get(id)` method just constructs a new `GobblinMetrics` instance with a given `id`. This `id` is used when constructing the `OutputStreamMetricReporter`

kadaan commented:
@sahilTakiar I'm running this locally and so far the issues seem to be fixed.  I still have metrics and jmx disabled, but I'll try it with them enabled later today.

sahilTakiar commented:
@kadaan awesome! good to know. I will work on rebasing this PR and addressing your comments sometime today.

sahilTakiar commented:
@kadaan updated
* Rebased the PR
* `ServiceBasedAppLauncher` can now run a configurable set of `Service`s specified by the config key `app.additional.services`
* Addressed other comments

sahilTakiar commented:
@zliu41 can you also take a look at this when you have a chance?

zliu41 commented on gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java:
```diff
@@ -0,0 +1,242 @@
+      LOG.warn("ApplicationLauncher has already started");
+      return;
+    } else {
+      this.hasStarted = true;
```
should set this.hasStopped = false?

zliu41 commented on gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java:
```diff
@@ -0,0 +1,242 @@
+   */
+  @Override
+  public void stop() throws ApplicationException {
+    if (this.hasStopped) {
```
should be `if (!this.hasStarted || this.hasStopped)`?

zliu41 commented on gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java:
```diff
@@ -0,0 +1,242 @@
+      LOG.warn("ApplicationLauncher has already stopped");
+      return;
+    } else {
+      this.hasStopped = true;
```
should set  hasStarted = false?

zliu41 replied:
actually why not just use a single boolean variable to denote started vs stopped.

sahilTakiar replied:
A Guava `Service` cannot be re-started, so the `ServiceBasedAppLauncher` shouldn't be able to restart either, which is why two variables are necessary.

zliu41 commented on gobblin-azkaban/src/main/java/gobblin/azkaban/AzkabanJobLauncher.java:
```diff
@@ -49,7 +58,7 @@
  * @author Yinan Li
  */
-public class AzkabanJobLauncher extends AbstractJob {
+public class AzkabanJobLauncher extends AbstractJob implements ApplicationLauncher, JobLauncher {
```
why does this class need to implement `JobLauncher`? When should one call `launchJob` directly instead of `run`?

sahilTakiar replied:
Its mostly just to declare that `AzkabanJobLauncher` is a `JobLauncher` - that is, it is capable of launching jobs. In this specific scenario there wouldn't really be case where someone would directly call `launchJob`, I think it just makes the code easier to understand.

zliu41 replied:
It may be a little confusing. Because `job.launcher.type` can only be LOCAL, MAPREDUCE or YARN. It cannot be AzkabanJobLauncher. So technically it's not a job launcher, at least not in the same sense as LocalJobLauncher or MRJobLauncher.

sahilTakiar replied:
True, but I think its more like `CliLocalJobLauncher extends LocalJobLauncher` or `CliMRJobLauncher extends MRJobLauncher` - but I wanted to avoid using `extends` since many of these classes already `extend` some other class.

`JobLauncherFactory` only has two publicly available launch types, `LOCAL` and `MAPREDUCE` and provides implementations for both. I could add the `AzkabanJobLauncher` here, but it wouldn't provide any benefit.

I think `JobLauncherFactory` is useful for creating `JobLauncher`s in code, but that shouldn't prevent creation of `JobLauncher`s that are directly invoked using a `main` method.

zliu41 commented on gobblin-runtime/src/main/java/gobblin/runtime/app/ServiceBasedAppLauncher.java:
```diff
@@ -0,0 +1,242 @@
+    LOG.info("Starting the Gobblin application and all its associated Services");
+
+    // Start the application
+    this.serviceManager.startAsync();
```
should probably add `this.serviceManager.awaitHealthy()`

sahilTakiar replied:
Added

sahilTakiar commented:
@zliu41 updated and comments addressed

zliu41 commented:
LGTM

sahilTakiar commented:
@kadaan any other comments?

kadaan commented:
@sahilTakiar LGTM (metrics work)
